core/src/authentication.rs

```diff
@@ -63,42 +61,33 @@ impl Credentials {
             Ok(data)
         }
 
-        let encrypted_blob = base64::decode(encrypted_blob).unwrap();
-
-        let secret = {
-            let mut data = [0u8; 20];
-            let mut h = crypto::sha1::Sha1::new();
-            h.input(device_id.as_bytes());
-            h.result(&mut data);
-            data
-        };
+        let secret = Sha1::digest(device_id.as_bytes());
 
         let key = {
-            let mut data = [0u8; 24];
-            let mut mac = Hmac::new(Sha1::new(), &secret);
-            pbkdf2(&mut mac, username.as_bytes(), 0x100, &mut data[0..20]);
-
-            let mut hash = Sha1::new();
-            hash.input(&data[0..20]);
-            hash.result(&mut data[0..20]);
-            BigEndian::write_u32(&mut data[20..], 20);
-            data
+            let mut key = [0u8; 24];
+            pbkdf2::<Hmac<Sha1>>(&secret, username.as_bytes(), 0x100, &mut key[0..20]);
+
+            let hash = &Sha1::digest(&key[..20]);
+            key[..20].copy_from_slice(hash);
+            BigEndian::write_u32(&mut key[20..], 20);
+            key
         };
 
+        // decrypt data using ECB mode without padding
         let blob = {
-            // Anyone know what this block mode is ?
-            let mut data = vec![0u8; encrypted_blob.len()];
-            let mut cipher =
-                aes::ecb_decryptor(aes::KeySize::KeySize192, &key, crypto::blockmodes::NoPadding);
-            cipher
-                .decrypt(
-                    &mut crypto::buffer::RefReadBuffer::new(&encrypted_blob),
-                    &mut crypto::buffer::RefWriteBuffer::new(&mut data),
-                    true,
-                )
-                .unwrap();
-
-            let l = encrypted_blob.len();
+            use aes::block_cipher_trait::BlockCipher;
+            use aes::block_cipher_trait::generic_array::GenericArray;
+            use aes::block_cipher_trait::generic_array::typenum::Unsigned;
+
+            let mut data = base64::decode(encrypted_blob).unwrap();
+            let cipher = Aes192::new(GenericArray::from_slice(&key));
+            let block_size = cipher::BlockSize::to_usize();
+            assert_eq!(data % block_size, 0);
+            for chunk in data.chunks_mut(block_size) {
+                cipher.decode(GenericArray::from_slice_mut(chunk));
+            }
```

Comment:
I am not sure about this part. Can it be assumed that `data` here is always multiple of 16?

Reply:
it doesn't look like it. I just checked with my stored blob, where the hex blob was 286 characters

Reply:
Hm, this is quite strange. If ECB (or CBC) mode was used encrypted message length should be a multiple of block size. Do you have any test vectors for those encrypted blocks or implementations which can be used as a reference?

Reply:
Will have to check later. I don't believe so, I believe we just store the bllob as we receive it, no telling what parameters are used in it's creation.